hs/src/Reach/Simulator.hs

```diff
@@ -0,0 +1,477 @@
+  (PGE, [V_UInt lhs,V_UInt rhs]) -> return $ V_Bool $ (>=) lhs rhs
+  (PGT, [V_UInt lhs,V_UInt rhs]) -> return $ V_Bool $ (>) lhs rhs
+  (IF_THEN_ELSE, [V_Bool cond, cond_val, alt]) -> do
+    return $ if cond then cond_val else alt
+  (DIGEST_EQ,  [V_Digest lhs,V_Digest rhs]) -> return $ V_Bool $ (==) lhs rhs
+  (ADDRESS_EQ,  [V_Address lhs,V_Address rhs]) -> return $ V_Bool $ (==) lhs rhs
+  (TOKEN_EQ, [V_Token lhs,V_Token rhs]) -> return $ V_Bool $ (==) lhs rhs
+  -- TODO
+  -- (SELF_ADDRESS, _) -> undefined
+  (LSH, [V_UInt lhs,V_UInt rhs]) -> return $ V_UInt $ shiftL lhs (fromIntegral rhs)
+  (RSH, [V_UInt lhs,V_UInt rhs]) -> return $ V_UInt $ shiftR lhs (fromIntegral rhs)
+  (BAND, [V_UInt lhs,V_UInt rhs]) -> return $ V_UInt $ (.&.) lhs rhs
+  (BIOR, [V_UInt lhs,V_UInt rhs]) -> return $ V_UInt $ (.|.) lhs rhs
+  (BXOR, [V_UInt lhs,V_UInt rhs]) -> return $ V_UInt $ xor lhs rhs
+  _ -> impossible "unexpected error"
+
+instance Interp DLArg where
+  interp = \case
+    DLA_Var dlvar -> do
+      g <- globalGet
+      let st = e_store g
+      return $ st M.! dlvar
+    -- TODO: fake constants per connector
+    DLA_Constant _dlconst -> return $ V_UInt $ toInteger (maxBound :: Int)
+    DLA_Literal dllit -> interp dllit
+    -- TODO: handle interact case: needs participant frontend state
+    DLA_Interact _slpart _string _dltype -> undefined
+
+instance Interp DLLiteral where
+  interp = \case
+    DLL_Null -> return $ V_Null
+    DLL_Bool bool -> return $ V_Bool bool
+    DLL_Int _at int -> return $ V_UInt int
+
+instance Interp DLLargeArg where
+  interp = \case
+    DLLA_Array _dltype dlargs -> V_Array <$> mapM interp dlargs
+    DLLA_Tuple dlargs -> V_Tuple <$> mapM interp dlargs
+    DLLA_Obj strs_to_dlargs -> V_Object <$> mapM interp strs_to_dlargs
+    DLLA_Data _map_slvars_to_dltypes string dlarg -> do
+      evd_arg <- interp $ dlarg
+      return $ V_Data string evd_arg
+    DLLA_Struct assoc_slvars_dlargs -> do
+      evd_args <- mapM (\arg -> interp arg) $ M.fromList assoc_slvars_dlargs
+      return $ V_Struct $ M.toList evd_args
+    DLLA_Bytes _ -> undefined
+
+instance Interp DLExpr where
+  interp = \case
+    DLE_Arg _at dlarg -> interp dlarg
+    DLE_LArg _at dllargearg -> interp dllargearg
+    DLE_Impossible at _int err -> expect_thrown at err
+    DLE_PrimOp _at primop dlargs -> do
+      evd_args <- mapM interp dlargs
+      interpPrim (primop,evd_args)
+    DLE_ArrayRef _at dlarg1 dlarg2 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      case (ev1,ev2) of
+        (V_Array arr, V_UInt n) -> return $ arr !! (fromIntegral n)
+        _ -> impossible "expression interpreter"
+    DLE_ArraySet _at dlarg1 dlarg2 dlarg3 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      ev3 <- interp dlarg3
+      case (ev1,ev2) of
+        (V_Array arr, V_UInt n) -> do
+          let n' = fromIntegral n
+          return $ V_Array $ take n' arr ++ [ev3] ++ drop (n' + 1) arr
```

Comment:
I think we've written this like five times in the code base. You should find those cases and make a little helper function that they and you can all call